inception-concept-linking/src/main/java/de/tudarmstadt/ukp/inception/conceptlinking/util/QueryUtil.java

```diff
@@ -67,24 +66,106 @@
 
     private static final String WIKIMEDIA_NAVIGATIONAL_TEMPLATE = "e:Q11753321";
-    /**
-     *
-     * @param tokens the words spanned by the mention
-     * @param limit maximum number of results
-     * @return a query to retrieve candidate entities
-     */
-    public static TupleQuery generateCandidateQuery(RepositoryConnection conn, String tokens,
-        int limit, IRI aDescriptionIri)
+
+    private static String getExactMatchingQueryPart(String aString)
     {
+        return String.join("\n",
+            "    SELECT DISTINCT ?e2 ?description WHERE",
+            "    {",
+            "     VALUES ?labelpredicate {rdfs:label skos:altLabel}",
+            "      {",
+            "        ?e2 ?labelpredicate ?" + aString + " @en .",
+            "        OPTIONAL",
+            "        {",
+            "          ?e2 ?descriptionIri ?description.",
+            "          FILTER ( lang(?description) = \"en\" )",
+            "        }",
+            "      }",
+            "    }");
+    }
+
+    public static TupleQuery generateCandidateExactQuery(RepositoryConnection conn,
+        String aTypedString, String aMention, IRI aDescriptionIri)
+    {
+        // Matching user input exactly
+        String exactMatchingTypedString = getExactMatchingQueryPart("exactTyped");
+
+        // Usually users are lazy and don't capitalize words,
+        // but then the exact matching wouldn't find anything.
+        String exactMatchingTypedStringCapitalized = getExactMatchingQueryPart("exactTypedCapitalized");
+
+        // Do the same for the mention
+        String exactMatchingMention = getExactMatchingQueryPart("exactMention");
+
+        String exactMatchingMentionCapitalized = getExactMatchingQueryPart("exactMentionCapitalized");
+
         String query = String.join("\n",
             "DEFINE input:inference 'instances'",
             SPARQL_PREFIX,
             "SELECT DISTINCT ?e2 ?altLabel ?label ?description WHERE",
             "{",
             "  {",
+            exactMatchingTypedString,
+            "  } ",
+            "  UNION",
+            "  {",
+            exactMatchingTypedStringCapitalized,
+            "  }",
+            "  UNION",
+            "  {",
+            exactMatchingMention,
+            "  }",
+            "  UNION",
+            "  {",
+            exactMatchingMentionCapitalized,
+            "  }",
+            "  FILTER EXISTS { ?e2 ?p ?v }",
+            "  FILTER NOT EXISTS ",
+            "  {",
+            "    VALUES ?topic {" + String.join(" ", WIKIMEDIA_INTERNAL,
+                WIKIMEDIA_PROJECT_PAGE, WIKIMEDIA_CATEGORY, WIKIMEDIA_DISAMBIGUATION_PAGE,
+                WIKIMEDIA_LIST_ARTICLE, WIKIMEDIA_TEMPLATE, WIKIMEDIA_NEWS_ARTICLE,
+                WIKIMEDIA_NAVIGATIONAL_TEMPLATE) +
+                "}",
+            "    ?e2 rdf:type ?topic",
+            "  }",
+            "  ?e2 rdfs:label ?label.",
+            "  FILTER ( lang(?label) = \"en\" )",
+            "}");
```

Comment:
Is there no limit here?

Reply:
I am assuming that there are not many objects matching exactly.